File: LibComArduino.py
#\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/
# FILE: libcomarduino.py
# VERSION: 0.1.00
# PURPOSE: Easy communication Host <-> Arduino
# CREATED BY: Arnaud D. LINA
# DATE: June 2014
#/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\

import os
import serial
import time

DEFAULT_DEVICE_NAME         = "Arduino_Board"
DEFAULT_COM_PORT            = "/dev/tty.usbmodem1451"
DEFAULT_VERBOSE             = True
DEFAULT_TIMEOUT_ACKNOWLEDGE = 5   # seconds
DEFAULT_TIMEOUT_OPEN        = 120 # seconds
DEFAULT_TIMEOUT_READ        = 1   # seconds

#---------------------------------------------------------------------
class ComArduino:
   
   #---------------------------------------------------------------------
   def __init__(self, com_name= DEFAULT_DEVICE_NAME, com_port=DEFAULT_COM_PORT):
      
      self.com_port             = com_port
      self.com_name             = com_name
      self.com_serial           = None
      self.verbose              = DEFAULT_VERBOSE
      self.timeout_acknowledge  = DEFAULT_TIMEOUT_ACKNOWLEDGE
      self.timeout_read         = DEFAULT_TIMEOUT_READ
      self.timeout_open         = DEFAULT_TIMEOUT_OPEN
      self.com_unit_speed       = None
      self.connected            = False
   
   #---------------------------------------------------------------------
   def Open(self, com_baudrate=9600):

      # Open serial port
      if(self.verbose==True):
         print("Opening the port...")
         
      self.com_serial = serial.Serial(self.com_port)

      # Set serial port properties
      self.com_serial.baudrate = com_baudrate
      self.com_serial.bytesize = serial.EIGHTBITS
      self.com_serial.parity   = serial.PARITY_NONE
      self.com_serial.stopbits = serial.STOPBITS_ONE
      self.com_serial.timeout  = 2
      self.com_serial.xonxoff  = 0
      self.com_serial.rtscts   = 0

      # Pause
      time.sleep(2.0)

      if(self.com_serial.isOpen()==False):
         if(self.verbose==True):
            print(" -Open fail...")
         return False

      if(self.verbose==True):
         print(" -Open pass...")
      
      # Check device identity
      if(self.verbose==True):
         print("Checking device identity...")
         
      self.Identify=None
      initime=time.time()
      while((self.__GetWord(self.Identify, 0)!=self.com_name) and ((time.time()-initime)<self.timeout_open)):
         self.Identify = self.ReadData()
         if(self.verbose==True):
            print(" -[Read|Ref::"+self.__GetWord(self.Identify, 0)+"|"+self.com_name+"]")

      # Measure communication Rx speed, then Acknowledge
      if(self.__GetWord(self.Identify,0)==self.com_name):

         if(self.verbose==True):
            print(" -Connection opened...")
         
         self.connected = True
         self.__EvalComSpeed();
         time.sleep(2.0);
         self.WriteData(104);

         if(self.verbose==True):
            self.Print()
         
         return True         
         
      if(self.verbose==True):
         print(" -Connection fail...")

      return False
   
   #---------------------------------------------------------------------
   def WriteData(self, data, acknowledge=False):
      
      if(self.com_serial.isOpen()):
         self.com_serial.write(str(data))
         
         while True:
            try:
               time.sleep(0.01)
               break
            except:
               pass
            
         time.sleep(0.1)
      
      
      if(acknowledge==True):
         data = ''
         initime=time.time()
         while((data!='104') and ((time.time()-initime)<self.timeout_acknowledge)):
            data=self.__GetWord(str(self.ReadData()), 0)
         return (data=='104')

      ### MODIFICATION: IF ACKNOWLEDGEMENT, WE SHOULD TRY TO WRITE & FLUSH UNITL WE GET ANSWER
      ### OR SLEEP LONGER TO SEE
      
      return True
   
   #---------------------------------------------------------------------
   def ReadData(self, acknowledge=False):

      initime=time.time()
      data=''
      while((data=='') and ((time.time()-initime)<self.timeout_read)):
         try:
            data=str(self.com_serial.readline())
         except:
               pass
 
      if((data!='') and (acknowledge)):
         self.WriteData(104)
         
      if(self.verbose==True):
         print("Read: "+data)

      return data

   #---------------------------------------------------------------------
   def __EvalComSpeed(self):

      initime=time.time()

      for x in range(0, 10):
         self.WriteData(1)

      self.com_unit_speed = (time.time()-initime)/10.0
      self.com_serial.flush()

      if(self.verbose==True):
         print("ComSpeed: {0:.3f}msec".format(self.com_unit_speed*1000))

   #---------------------------------------------------------------------
   def Print(self):

      if(self.com_serial and self.com_serial.isOpen() and self.connected):
         print("==========================================================")
         print "SERIAL COM INFO"
         print " - Status   : opened and connected"
         print " - Name     : "  + self.__GetWord(self.Identify, 0)
         print " - Port     : "  + str(self.com_port)
         print " - Baudrate : "  + str(self.com_serial.baudrate)
         print " - Bytesize : "  + str(self.com_serial.bytesize)
         print " - Parity   : "  + str(self.com_serial.parity)
         print " - Stopbits : "  + str(self.com_serial.stopbits)
         print " - Timeout  : "  + str(self.com_serial.timeout)
         print " - Xonxoff  : "  + str(self.com_serial.xonxoff)
         print " - Rtscts   : "  + str(self.com_serial.rtscts)
         print(" - Tx speed : {0:.3f}msec".format(self.com_unit_speed*1000))
         print("==========================================================")
      else:
         print("==========================================================")
         print "SERIAL COM INFO"
         print " - Status: closed/not_connected"
         print("==========================================================")
         
#---------------------------------------------------------------------
   def __GetWord(self, string, index):
      if(string!=None):
         if(len(string)>0):
            return string.split()[index]
      return ''

File: LibComArduino.cpp
//\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/
// FILE: libcomarduino.cpp
// VERSION: 0.1.00
// PURPOSE: Easy communication  Host <-> Arduino
// CREATED BY: Arnaud D. LINA
// DATE: June 2014
//\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\

#include "libcomarduino.h"

#define DEFAULT_TIMEOUT_OPEN              10 // seconds
#define DEFAULT_TIMEOUT_OPEN_ACKNOWLEDGE  30 // seconds
#define DEFAULT_TIMEOUT_ACKNOWLEDGE        2 // seconds

/////////////////////////////////////////////////////
//
// PUBLIC
//

ComArduino::ComArduino()
   {
   verbose                    = false;
   timeout_open               = DEFAULT_TIMEOUT_OPEN;
   timeout_open_acknowledge   = DEFAULT_TIMEOUT_OPEN_ACKNOWLEDGE;
   timeout_acknowledge        = DEFAULT_TIMEOUT_ACKNOWLEDGE;
   funclevel                  = 0;
   }

boolean ComArduino::Open(bool Acknowledge)
   {
   FuncObj obj(&funclevel, &verbose);
   return funcOpen(Acknowledge);
   }

boolean ComArduino::IsConnected()
   {
   FuncObj obj(&funclevel, &verbose);
   return funcIsConnected();
   }

boolean ComArduino::IsDataIn()
   {
   FuncObj obj(&funclevel, &verbose);
   return funcIsDataIn();
   }


boolean ComArduino::ReadInt(int*    InPtr,
                            boolean Acknowledge)
   {
   FuncObj obj(&funclevel, &verbose);
   
   if(InPtr==NULL) {Serial.println("ERROR 0");};
   
   return funcReadInt(InPtr, Acknowledge);

   }

boolean ComArduino::ReadMultipleInt(int*    InPtr,
                                    int     Maxsize,
                                    int&    Realsize,
                                    boolean Acknowledge)
   {
   FuncObj obj(&funclevel, &verbose);
   
   if(InPtr==NULL)   {Serial.println("ERROR 1");};
   if(Maxsize<0)     {Serial.println("ERROR 2");};
   
   return funcReadMultipleInt(InPtr, Maxsize, Realsize, Acknowledge);
   }


boolean ComArduino::WriteInt(int     Out,
                             boolean Acknowledge)
   {
   FuncObj obj(&funclevel, &verbose);
   return funcWriteInt(Out, Acknowledge);
   }

boolean ComArduino::WriteMultipleInt(int*    OutPtr,
                                     int     Size,
                                     boolean Acknowledge)
{
   FuncObj obj(&funclevel, &verbose);
   
   if(OutPtr==NULL)  {Serial.println("ERROR 3");};
   if(Size<0)        {Serial.println("ERROR 4");};
   
   return funcWriteMultipleInt(OutPtr, Size, Acknowledge);
}

boolean ComArduino::WriteStr(String    Out,
                             boolean   Acknowledge)
{
   FuncObj obj(&funclevel, &verbose);
   return funcWriteStr(Out, Acknowledge);
}

/////////////////////////////////////////////////////
//
// PRIVATE
//

boolean ComArduino::funcOpen(bool Acknowledge)
   {
   Serial.begin(9600);
  
   int Inc=timeout_open;
   while(!Serial && Inc--)
      delay(1000);
  
   Dump("[[START:Open(");
   if(Serial)
      {
      Serial.flush();
      boolean connected = false;
      int Inc=timeout_open_acknowledge;
      while((Inc--)&&(!connected))
         {
         Serial.flush();
         connected = WriteStr(String(ARDUINO_ID), Acknowledge);
         delay(1000-DEFAULT_TIMEOUT_ACKNOWLEDGE);
         }
      Serial.flush();
      
      // !! here add read/write speed average.
      
      Dump("Done)");
      return connected;
      }
   else
      {
      Dump("Fail)");
      return false;
      }
   Dump(")");
   }

boolean ComArduino::funcIsConnected()
   {
   Dump("IsConnected(");
   
   if(Serial)
      {
      Dump("Yes)");
      return true;
      }
   else
      {
      Dump("No)");
      return false;
      }
   Dump(")");
   }

boolean ComArduino::funcIsDataIn()
   {
   Dump("IsDataIn(");
   
   if(!IsConnected()) {return false;}
   
   if(Serial.available())
      {
      Dump("Yes)");
      return true;
      }
   else
      {
      Dump("No)");
      return false;
      }
   Dump(")");
   }


boolean ComArduino::funcReadInt(int*    InPtr,
                                boolean Acknowledge)
   {
   Dump("ReadInt(");
   
   if(!IsConnected()) {Dump(")"); return false;}
   
   int Realsize=0;
   if(IsDataIn())
      {
      *InPtr = Serial.parseInt();
      Realsize+=1;
      delay(20);
      Dump("+");
      }
   
   WriteAcknowledge(Acknowledge, Realsize);
   Dump(")");
   return true;
   }

boolean ComArduino::funcReadMultipleInt(int*    InPtr,
                                        int     Maxsize,
                                        int&    Realsize,
                                        boolean Acknowledge)
   {
   Dump("ReadMultipleInt(");
   
   if(!IsConnected()) {Dump(")"); return false;}
   
   Realsize=0;
   while((Realsize<Maxsize) && Serial.available())
      {
      *(InPtr+Realsize) = Serial.parseInt();
      Realsize+=1;
      Dump("+");
      }
   
   WriteAcknowledge(Acknowledge, Realsize);
   Dump(")");
   return true;
   }


boolean ComArduino::funcWriteInt(int     Out,
                                 boolean Acknowledge)
   {
   Dump("WriteInt(");
   
   if(!IsConnected()) {Dump(")"); return false;}
   
   Serial.println(Out);
   delay(20);
   
   boolean check = ReadAcknowledge(Acknowledge);
   Dump(")");
   return check;
   }

boolean ComArduino::funcWriteMultipleInt(int*    OutPtr,
                                         int     Size,
                                         boolean Acknowledge)
{
   Dump("WriteMultipleInt(");
   
   if(!IsConnected()) {Dump(")"); return false;}
   
   String Output;
   for(int ii=0; ii<(Size-1);ii++)
      {
      Output+=*(OutPtr+ii);
      Output+=" ";
      }
   Output+=*(OutPtr+(Size-1));
   
   Serial.println(Output);
   delay(20);
   
   boolean check = ReadAcknowledge(Acknowledge);
   Dump(")");
   return check;
}

boolean ComArduino::funcWriteStr(String  Out,
                                 boolean Acknowledge)
{
   Dump("WriteStr(");
   
   if(!IsConnected()) {Dump(")"); return false;}
   
   Serial.println(Out);
   delay(20);
   
   boolean check = ReadAcknowledge(Acknowledge);
   Dump(")");
   return check;
}

/////////////////////////////////////////////////////
//
// PRIVATE UTIL
//
void ComArduino::Dump(char* text)
   {
   if(verbose)
      {
      Serial.print(text);
      delay(20);
      }
   }

void ComArduino::WriteAcknowledge(boolean doit,
                                  int     value)
   {
   Dump("[WriteAcknowledge(");
   if(!doit) {Dump(")]"); return;}
   
   String textout = String("104,");
   textout+=value;
   textout+="\n";
   Serial.print(textout);
   delay(20);
   Dump(")]");
   }

boolean ComArduino::ReadAcknowledge(boolean doit)
   {
   Dump("[ReadAcknowledge(");
   if(!doit) {Dump(")]"); return true;}
   
   int In=0;
   int Inc=timeout_acknowledge;
   while(Inc--)
      {
      In = Serial.parseInt();
      delay(20);
      
      if(In==104)
         {
         Dump("done)]");
         return true;
         }
      delay(1000);
      }
   Dump("timeout)]");
   return false;
   }


//
// END OF FILE
//
